### PR TITLE
feat: split js_image_layer into fine grained layers

### PR DESCRIPTION
**Pull request by gregmagolan**
Resolves https://github.com/aspect-build/rules_js/issues/1641.

This change splits the output of js_image_layer into 5 layers:

1. `node` layer contains Node.js and the node fs patches
2. `package_store_3p` layers containers all 3p deps in the `.aspect_rules_js` package store
3. `package_store_1p` layers containers all 1p deps in the `.aspect_rules_js` package store
4. `node_modules` layer contains all `node_modules/foo` symlinks into the packages tore
5. `app` layer contains everything that doesn't fall into any of the above layers

### CODE REVIEW

**Comment by thesayyn** [comment]
I think we should put node and node_modules into same layer as those are less likely to change.

**Comment by thesayyn** [comment]
hmm how about; `node_modules_links  + package_store_3p + node ` they all go into the same layer, as they won't change frequently?

**Comment by gregmagolan** [comment]
> hmm how about; `node_modules_links + package_store_3p + node ` they all go into the same layer, as they won't change frequently?

They could; node_modules_links include 1p links but won't change as much as the actual contents of the 1p links. Is there a downside to more fine grained layers?

**Comment by thesayyn** [comment]
Other than network round trip not really, more the better, but i am afraid that we are splitting it too much that it will be confusing.

**Comment by gregmagolan** [comment]
> Other than network round trip not really, more the better, but i am afraid that we are splitting it too much that it will be confusing.

Its transparent for users of rules_oci tho right since `tars = [":layers",],` in `oci_image`. Just the rules_docker example has them listed explicitly and we don't care about rules_docker ease of use since that rule set is abandonware now.

**Comment by thesayyn** [comment]
Makes sense, i guess this is fine, fine grained layering will give users more power.

**Comment by matthewjh** [comment]
Completely agree with fine-grained layering along these lines - it should save a lot of bandwidth and time for orgs that have 100+ services.

Thanks @gregmagolan!

File: js/private/js_image_layer.bzl [hunk 1]
```diff
@@ -209,6 +279,9 @@ def _runfile_path(ctx, file, runfiles_dir):
     return paths.join(runfiles_dir, to_rlocation_path(ctx, file))
 
 def _build_layer(ctx, type, all_entries_json, entries, inputs):
+    if not entries:
+        return None
```
Comment (gregmagolan) [inline js/private/js_image_layer.bzl:283]: short-circuit and don't create the layer if it has no entries

**Comment by thesayyn** [comment]
I think this PR has a bad rebase, @gregmagolan could you fix that before i can take a look?

**Comment by gregmagolan** [comment]
> I think this PR has a bad rebase, @gregmagolan could you fix that before i can take a look?

Ahh yes, the base branch was rebased which messed up the commits. All clean again now.

**Review by thesayyn — APPROVED** [review]

**Pull request merged**